### PR TITLE
Adds support for instrumentation using ActiveSupport notifications

### PR DESCRIPTION
**Pull request by gregbeech**
Makes it easier to monitor event publishing, and also could be used to
provide in-process handling of events.

### CODE REVIEW

**Comment by gregbeech** [comment]
@jphastings @mezis fancy a look at this?

File: spec/rooted/publisher_spec.rb [hunk 1]
```diff
@@ -15,20 +15,52 @@ def as_json
 
   let(:adapter) { instance_double(Rooted::Adapters::Memory) }
   let(:publisher) { Rooted::Publisher.new(adapter: adapter) }
+  let(:resource) { TestObject.new }
+  let(:metadata) { { foo: 'bar' } }
+  let(:perform) { publisher.public_send(verb, resource, metadata) }
 
   %i(created updated deleted).each do |verb|
     describe "##{verb}" do
+      let(:verb) { verb }
```
Comment (jphastings) [inline spec/rooted/publisher_spec.rb]: It's probably not great to shadow like this, plus I think `verb` should still be available inside the `it` blocks without the `let` definition.

**Comment by jphastings** [comment]
LGTM, one small comment about variable shadowing, otherwise looks good!

File: spec/rooted/publisher_spec.rb (hunk 1 above)
Comment (gregbeech) [inline spec/rooted/publisher_spec.rb]: It's not variable shadowing as this code runs at different times; the loop runs when the class definition is built and then the `let` runs when the class is executed (at which point `verb` is not a variable, it's baked into the definition). I could have substituted the loop variable directly into the code, but when generating tests I tend to prefer to move the variables into `let`s immediately to make it clearer what's happening.

**Pull request merged**